pkg/resources/kafka/poddisruptionbudget.go

```diff
@@ -0,0 +1,57 @@
+// Copyright © 2019 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafka
+
+import (
+	"fmt"
+
+	"github.com/banzaicloud/kafka-operator/pkg/resources/templates"
+	"github.com/banzaicloud/kafka-operator/pkg/util"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (r *Reconciler) poddisruptionbudget() runtime.Object {
+	minAvailable := r.computeMinAvailable()
+
+	return &policyv1beta1.PodDisruptionBudget{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "PodDisruptionBudget",
+			APIVersion: "policy/v1beta1",
+		},
+		ObjectMeta: templates.ObjectMetaWithAnnotations(
+			fmt.Sprintf("%s-pdb", r.KafkaCluster.Name),
+			util.MergeLabels(LabelsForKafka(r.KafkaCluster.Name), r.KafkaCluster.Labels),
+			r.KafkaCluster.Spec.ListenersConfig.GetServiceAnnotations(),
+			r.KafkaCluster,
+		),
+		Spec: policyv1beta1.PodDisruptionBudgetSpec{
+			MinAvailable: &minAvailable,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: LabelsForKafka(r.KafkaCluster.Name),
+			},
+		},
+	}
+
+}
+
+// Calculate maxUnavailable as max between brokerCount - 1 (so we only allow 1 broker to be disrupted)
+// and 1 (to cover for 1 broker clusters)
+func (r *Reconciler) computeMinAvailable() intstr.IntOrString {
```

Comment:
@cojan minAvailable needs to be configurable. Value 1 works for cases when oneBrokerPerNode is configured but not when this is disabled. In the latter case there may be multiple brokers per node and a value 1 of minAvailable will imped the k8s cluster upgrade if there are multiple broker pods scheduled on the same node.

My proposal would be to put this pbd under a feature flag and let the user configure the actual value of minAvailable

Reply:
@amuraru I agree that minAvailable might be worth configuring to some degree, I actually was planning on asking for some guidance where this configuration should be stay.

In the latter case that you describe, the PDB will act the same as for the "oneBrokerPerNode", the only difference would be in the time required for the node to be drained (as 2 or more brokers need to be moved before finishing the node drain). In the case of multiple brokers on a single node, it's even more critical in my opinion.

Agreed to put this under a feature flag though.

If I'm missing anything please let me know.

Reply:
I take my comment back - it seems that that k8s descheduler/drain can work with 1  minUnvailable replica and multiple kafka pods on the same node.
I don't think that making this configurable unless the banzaicloud maintainers feel so

Reply:
I like the idea of putting this under a flag and make it configurable. Also what do you think instead of using a number we should use percentage. Using percentage we can support scenarios where the cluster can tolerate more than one broker outage e.g.: when all your topics has three replication factors.
I know we cannot use percentage on the PDB side but the operator can calculate the number based on the percentage specified in the KafkaCluster spec.

Reply:
All for putting it under a feature flag in the KafkaCluster spec, was thinking about:
```yaml
disruptionBudget:
 create: true
 budget: "33%"
```
Let me know what are your thoughts.

I was also wondering if we could support both percentage and static number, wdyt?

I was looking at a case with replication factor 2 and multi AZ. When switching to the next VM from an AZ to another, there is a chance to have 2 brokers down at the same time and both of them to hold one of the replicas. I know it's a corner-case  but I would like to have the flexibility of using a number as well.

Reply:
I like it. Yeah it is reasonable to support both percentage and static number.

Reply:
Had a bit of time today and added the extra work to implement the above, please have another look.